### PR TITLE
HDDS-7943. Link Datanode WebUIs from the Node Status table on SCM WebUI

### PR DESCRIPTION
**Pull request by debiswal**
## What changes were proposed in this pull request?

1. Reprsent DataNode Link on SCM UI with HTTP & HTTPS Port
2. Three  cases for every DN and SCM as below
a. HTTP_ONLY -> the service does serve its webUI via HTTP
b. HTTPS_ONLY -> the service does serve its webUI via HTTPS
c. HTTP_AND_HTTPS -> the service does serve its webUI via HTTP and HTTPS both based on protocol scheme

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7943

## How was this patch tested?
Manually with JMX Response

**Response Object:**
```
[ {
"key" : "ozonesecure_datanode_3.ozonesecure_default",
"value" : [ {
"key" : "HTTPS",
"value" : "9883"
}, {
"key" : "OPSTATE",
"value" : "IN_SERVICE"
}, {
"key" : "COMSTATE",
"value" : "HEALTHY"
}, {
"key" : "HTTP",
"value" : "9882"
} ]
}, {
"key" : "ozonesecure_datanode_2.ozonesecure_default",
"value" : [ {
"key" : "HTTPS",
"value" : "9883"
}, {
"key" : "OPSTATE",
"value" : "IN_SERVICE"
}, {
"key" : "COMSTATE",
"value" : "HEALTHY"
}, {
"key" : "HTTP",
"value" : "9882"
} ]
}, {
"key" : "ozonesecure_datanode_1.ozonesecure_default",
"value" : [ {
"key" : "HTTPS",
"value" : "9883"
}, {
"key" : "OPSTATE",
"value" : "IN_SERVICE"
}, {
"key" : "COMSTATE",
"value" : "HEALTHY"
}, {
"key" : "HTTP",
"value" : "9882"
} ]
} ]
```

### CODE REVIEW

**Comment by kerneltime** [comment]
cc @devmadhuu can you please take a look?

**Comment by dombizita** [comment]
hi @debiswal, thanks for taking this up, please take a look at the CI, the `findbugs` failed with this message:
```
M C NP: Possible null pointer dereference of health in org.apache.hadoop.hdds.scm.node.SCMNodeManager.getNodeStatusInfo()  Dereferenced at SCMNodeManager.java:[line 1084]
M C NP: Possible null pointer dereference of nodestatus in org.apache.hadoop.hdds.scm.node.SCMNodeManager.getNodeStatusInfo()  Dereferenced at SCMNodeManager.java:[line 1076]
M C NP: Possible null pointer dereference of operationalState in org.apache.hadoop.hdds.scm.node.SCMNodeManager.getNodeStatusInfo()  Dereferenced at SCMNodeManager.java:[line 1083]
```

**Comment by ArafatKhan2198** [comment]
Hi @debiswal I attempted to access the SCM-UI by creating an HA-Cluster on my local Docker setup using three datanodes. However, I was unable to access any of the web UIs for the datanodes, and discovered that all three were listening on port 9882, with identical DataNode UI links. Please let me if there is anything I'm missing.

image
image

**Comment by debiswal** [comment]
> Hi @debiswal I attempted to access the SCM-UI by creating an HA-Cluster on my local Docker setup using three datanodes. However, I was unable to access any of the web UIs for the datanodes, and discovered that all three were listening on port 9882, with identical DataNode UI links. Please let me if there is anything I'm missing.
> 
> image
> 
> image

Datanode  webuis are  available inside of their own containers so when we are trying to access DN outside of their own containers it will not be available.

File: hadoop-hdds/framework/src/main/resources/webapps/static/ozone.css [hunk 1]
```diff
@@ -53,7 +53,36 @@ body {
 .sortorder.reverse:after {
     content: '\25bc';   // BLACK DOWN-POINTING TRIANGLE
 }
-
+.sorting:after {
```
Comment (fapifta) [inline hadoop-hdds/framework/src/main/resources/webapps/static/ozone.css:56]: These changes, and a couple of related changes seem to be something that does not belong to this JIRA, and fixing something that is not mentioned here.
It is great to fix this thing, but for clarity I think anything that is related to how ordering elements are being rendered should be in a separate JIRA describing the problem, and then fixed in a separate PR, so later on we can track and understand what has happened and why, and we can more easily find the track records of a particular problem.

If I misunderstand it, and this change is required to deliver the DataNode ports from the backend to the frontend, and then show DataNode WebUI URLs properly, then please correct me, if I am right, please file a separate JIRA for this and add the improvement for the sorting UI and logic separately from adding the URLs.

File: hadoop-hdds/server-scm/src/main/resources/webapps/scm/scm.js
Comment (fapifta) [inline hadoop-hdds/server-scm/src/main/resources/webapps/scm/scm.js:1]: Indentation of this file is pretty much messed up, please fix it. Please keep the 4 spaces as default indentation as with the rest of the file.

File: hadoop-hdds/server-scm/src/main/resources/webapps/scm/scm-overview.html [hunk 2]
```diff
@@ -45,23 +45,24 @@ <h2>Node Status</h2>
...
     <thead>
-        <tr>
-            <th ng-click = "columnSort('hostname')" class="nodeStausInfo"><span ng-class = "{'sorting' : (columnName != 'hostname'), 'sortasc' : (columnName == 'hostname' && !reverse),
-                                        'sortdesc':(columnName == 'hostname' && reverse)}">HostName</span></th>
-            <th ng-click = "columnSort('opstate')" class="nodeStausInfo" ><span ng-class="{'sorting' : (columnName != 'opstate'), 'sortasc' : (columnName == 'opstate' && !reverse),
-                                        'sortdesc':(columnName == 'opstate' && reverse)}">Operational State</span></th>
-            <th ng-click = "columnSort('comstate')" class="nodeStausInfo">  <span ng-class="{'sorting' : (columnName != 'comstate'), 'sortasc' : (columnName == 'comstate' && !reverse),
-                                        'sortdesc':(columnName == 'comstate' && reverse)}">Commisioned State</span> </th>
-        </tr>
+    <tr>
```
Comment (fapifta) [inline hadoop-hdds/server-scm/src/main/resources/webapps/scm/scm-overview.html:48]: Here again, I am not sure what is causing the indentation change, probably 4 spaces exchanged to tabs, or vice versa, but we should stick to keep 4 spaces as basic indentation as it is in the rest of the file.
Also these changes seem to be related again to fixing things with sorting, so these as well should go under a separate JIRA in a separate PR.

File: hadoop-hdds/server-scm/src/main/resources/webapps/scm/scm-overview.html [hunk 3]
```diff
@@ -45,23 +45,24 @@ <h2>Node Status</h2>
...
+    </tr>
     </thead>
     <tbody>
         <tr ng-repeat="typestat in nodeStatus|filter:search|orderBy:columnName:reverse">
-            <td>{{typestat.hostname}}</td>
+            <td><a href="{{typestat.portval.toLowerCase()}}://{{typestat.hostname}}:{{typestat.portno}}" target="_blank">{{typestat.hostname}}</a></td>
             <td>{{typestat.opstate}}</td>
             <td>{{typestat.comstate}}</td>
-        </tr>
+       </tr>
```
Comment (fapifta) [inline hadoop-hdds/server-scm/src/main/resources/webapps/scm/scm-overview.html:62]: Incorrect indentation change, please keep 4 spaces as for the rest of the file.

**Review by fapifta — CHANGES_REQUESTED** [review]
Hi @debiswal thank you for working on this piece. Please find my review comments inline, mostly it is about indentation, and also about an ask to separate away the sorting related fixes that were -I believe - mixed in into this commit, if I am wrong on that, please do not hesitate to correct me, but in this case please detail a bit in the PR description why the sorting related changes are here.

File: hadoop-hdds/framework/src/main/resources/webapps/static/ozone.css (hunk 1 above)
Comment (debiswal) [reply hadoop-hdds/framework/src/main/resources/webapps/static/ozone.css]: Removed Sorting related Data Rendering changes & will be tracked in a newly created  JIRA : HDDS-8465

File: hadoop-hdds/server-scm/src/main/resources/webapps/scm/scm.js
Comment (debiswal) [reply hadoop-hdds/server-scm/src/main/resources/webapps/scm/scm.js:1]: Addressed

**Comment by debiswal** [comment]
> Hi @debiswal thank you for working on this piece. Please find my review comments inline, mostly it is about indentation, and also about an ask to separate away the sorting related fixes that were -I believe - mixed in into this commit, if I am wrong on that, please do not hesitate to correct me, but in this case please detail a bit in the PR description why the sorting related changes are here.

New JIRA Created for sorting related fix :  https://issues.apache.org/jira/browse/HDDS-8465

**Comment by fapifta** [comment]
Hi @debiswal thank you for opening the new JIRA, and to update this patch.
As I see the indentation in scm.js is still not correct here and there, please take one more look at it.

**Review by fapifta — APPROVED** [review]
LGTM, thank you for addressing the indentation issues. LGTM.
